### PR TITLE
Feature/dependency observer

### PR DESCRIPTION
**Pull request by HenrikThoroe**
This PR updates the compiler to be able to recompile every file which depends on a changed header file.

### CODE REVIEW

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/dependency_graph.dart:11]: Why not just make it private or readonly with a getter?

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/dependency_graph.dart:7]: Missing newline

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/dependency_graph.dart]: I think they are called 'transitive' dependencies, maybe change the name? I actually only know that because I looked into auto generating the licenses notice for dependencies.

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/dependency_graph.dart:92]: I like this use of recursion :)

You may resolve this, just wanted to let you know

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/dependency_graph.dart]: Why not use `.singleWhere`? For the custom error?

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/dependency_graph.dart:148]: What is the id? Is it the filename - as some code above might indicate?

File: lib/features/build_server/cache/dependencies/find_dependencies.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/find_dependencies.dart]:
```suggestion
/// Find all direct dependencies in [file].
```

File: lib/features/build_server/cache/dependencies/find_dependencies.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/find_dependencies.dart:11]: The `$`at the end is a good idea. But this would not be able to handle:
```cpp
/* We should maybe
#include "A"
or
#include "B"
*/
```
Which might be ok, but 
```cpp
#include "A" /* some comment */
```
should not raise errors.

Removing comments might be worth an extra function.

File: lib/features/build_server/cache/dependencies/resolve_dependency.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/resolve_dependency.dart:4]: Why did you leave out the `as p` here?

File: lib/features/build_server/cache/dependencies/resolve_dependency.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/resolve_dependency.dart]: The `!isWithin(...)` is extremely crucial to security, well spotted and though through 👍🏻

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [inline lib/features/build_server/cache/project_cache.dart]:
```suggestion
  /// The directory where all cache files reside.
```

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [inline lib/features/build_server/cache/project_cache.dart:123]: Not necessarily this PR, but this method could use some comments :)

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [inline lib/features/build_server/cache/project_cache.dart:174]: Why do that here, this seems a little random, maybe at least document it?

Comment (Coronon) [reply lib/features/build_server/cache/project_cache.dart:174]: We need this for `_isCompilationCandidate`, right? Please add this to the docstring (that we update dependecies) and add a comment in the code :)

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [inline lib/features/build_server/cache/project_cache.dart]: Maybe: 'Determine whether [file] has to be compiled based on previous compilation, extension and dependencies.' would be more descriptive?

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [inline lib/features/build_server/cache/project_cache.dart]: Just realizing this now, we never use `inSync`. Should be safe to remove. Probably an artifact of the old `sync`.

File: lib/features/build_server/cache/dependencies/resolve_dependency.dart
Comment (Coronon) [inline lib/features/build_server/cache/dependencies/resolve_dependency.dart]: A bit misleading, because you used the term 'entryDir' for the root before

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [inline lib/features/build_server/cache/project_cache.dart]: See comment on `entryDirectory` on **resolve_dependency.dart**

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [inline lib/features/build_server/cache/project_cache.dart:26]: I really like how you added more documentation (to the whole file), way to go 🚀

File: lib/features/build_server/cache/state/source_file_state.dart
Comment (Coronon) [inline lib/features/build_server/cache/state/source_file_state.dart]: I really dont understand your documentation, sry. The code is clear though.

File: test/features/build_server/cache/project_cache_test.dart
Comment (Coronon) [inline test/features/build_server/cache/project_cache_test.dart]: I would move these helpers to the bottom, main (what a pun) content first

Comment (Coronon) [reply test/features/build_server/cache/project_cache_test.dart]:
```suggestion
/// Reads the `./example` directory and returns a list of all cpp related files.
```

File: test/features/build_server/cache/project_cache_test.dart
Comment (Coronon) [inline test/features/build_server/cache/project_cache_test.dart]:
```suggestion
/// If [changing] is not null, all files with the name will be
```

File: test/features/build_server/cache/project_cache_test.dart
Comment (Coronon) [inline test/features/build_server/cache/project_cache_test.dart:22]: Maybe a bit high for test runners? (thread count)

File: test/features/build_server/cache/project_cache_test.dart
Comment (Coronon) [inline test/features/build_server/cache/project_cache_test.dart]:
```suggestion
        'constexpr int seven = 42;',
```
;)

**Review by Coronon — CHANGES_REQUESTED** [review]
Interesting but well though through approach :) I really like how you put more emphasis on comments and tests 👍🏻 

Mostly polish

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (HenrikThoroe) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart:11]: Because private fields are incompatible with json_serializable. Or do you know different?

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (HenrikThoroe) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart]: I guess someone can call them transitive, but just speaking for me I think 'indirect' states the purpose as good and is much more accessable to most who are reading the code. 'Indirect dependencies' is pretty clear in it's meaning and exactly what they are. 
Or what are your thoughts about it?

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (HenrikThoroe) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart]: Yes, that was my intention. Do you think I should instead use `singleWhere`?

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (HenrikThoroe) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart:148]: Actually it's really just the id of the node and `_DependencyNode` is not the instance in the code which determines what to use as the id. That's the job of `DependencyGraph`. Finally it will be set to the 'id' of each `SourceFile`.

Would you prefer me to add an implementation hint aka `It is recommended to set this field to the unique id of the corresponding source file` or similar?

File: lib/features/build_server/cache/dependencies/resolve_dependency.dart
Comment (HenrikThoroe) [reply lib/features/build_server/cache/dependencies/resolve_dependency.dart:4]: Because there are no naming conflicts

File: lib/features/build_server/cache/project_cache.dart
Comment (HenrikThoroe) [reply lib/features/build_server/cache/project_cache.dart:123]: Understandable, I will open another PR sometime soon, to add more documentation and maybe test coverage when needed

File: lib/features/build_server/cache/state/source_file_state.dart
Comment (HenrikThoroe) [reply lib/features/build_server/cache/state/source_file_state.dart]: Me neither, bro. Me neither...

File: test/features/build_server/cache/project_cache_test.dart
Comment (HenrikThoroe) [reply test/features/build_server/cache/project_cache_test.dart]: I prefer to put helpers at the top but I don't care enough to not resolve this :)

File: test/features/build_server/cache/project_cache_test.dart
Comment (HenrikThoroe) [reply test/features/build_server/cache/project_cache_test.dart:22]: I'd like to keep it at 4 to be able to test multi threaded compilation properly and because the example project is not meant to grow very big. It should not make any difference how many threads we are using.

File: test/features/build_server/cache/project_cache_test.dart
Comment (HenrikThoroe) [reply test/features/build_server/cache/project_cache_test.dart]: 0x2A

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart]: `singleWhere` would throw a StateError, which is quite descriptive, probably cleaner tbh. Decision is up to you.

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart:148]: Yes, such a hint would be great because it seems a bit arbitrary atm.

File: lib/features/build_server/cache/dependencies/resolve_dependency.dart
Comment (Coronon) [reply lib/features/build_server/cache/dependencies/resolve_dependency.dart:4]: While Im not a big fan of namespace pollution, it’s Dart style -> ok

File: lib/features/build_server/cache/project_cache.dart
Comment (Coronon) [reply lib/features/build_server/cache/project_cache.dart:123]: Well see about the coverage after this PR is merged ;)

File: test/features/build_server/cache/project_cache_test.dart
Comment (Coronon) [reply test/features/build_server/cache/project_cache_test.dart:22]: Sounds reasonable

File: test/features/build_server/cache/project_cache_test.dart
Comment (Coronon) [reply test/features/build_server/cache/project_cache_test.dart]: (-80538738812075974)^3 + 80435758145817515^3 + 12602123297335631^3

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart]: Indirect is probably easier to understand for most people. But ‘transitive’ is what they are actually called. When you look into the publock, node package management or just google ‘transitive’, the results are way higher quality.

File: lib/features/build_server/cache/dependencies/dependency_graph.dart
Comment (Coronon) [reply lib/features/build_server/cache/dependencies/dependency_graph.dart:11]: I just looked that up and its true, which really surprises me I have to admit. Maybe we should have a look at an alternative after this PR, e.g.: https://github.com/k-paxian/dart-json-mapper

**Review by Coronon — APPROVED** [review]

**Pull request merged**
